README.md
# ssm-provisioner

A script that provides the ability to provision AWS instances via SSM.

[Version v0.0.4](https://gitlab.com/zdzielinski/ssm-provisioner/-/tags/v0.0.4) - [Main GitLab Repository](https://gitlab.com/zdzielinski/ssm-provisioner) - [Secondary GitHub Mirror](https://github.com/zdzielinski/ssm-provisioner).

**NOTE**: This repository is currently unstable and a work in progress.

**NOTE**: Any commits, tags, or other content before release `v1.0.0` may be deleted or changed.

## Reasoning

AWS SSM is a secure alternative to SSH for general command execution and remote shell capability.

Unfortunately however, there is no standardized upstream SSM provisioner within the Terraform ecosystem.

This module comes into play as a simple, configurable option for SSM provisioning of AWS instances.

Instead of opening your instance up and provisioning over SSH, why not give SSM a try?

## Requirements

The following packages must be installed:

* [OpenSSH](https://www.openssh.com/)
* [Terraform](https://www.terraform.io/)
* [AWS CLI](https://aws.amazon.com/cli/)
* [SSM Plugin](https://docs.aws.amazon.com/systems-manager/latest/userguide/session-manager-working-with-install-plugin.html)

## Usage

The meat and potatoes of this project lives within the [ssm-provisioner.sh](./ssm-provisioner.sh) script.

One method of using this script is to download it into your module via the latest release tag:

```bash
curl -sO https://gitlab.com/zdzielinski/ssm-provisioner/-/raw/v0.0.4/ssm-provisioner.sh
```

Generally, this script would be leveraged within the provisioner of a `null_resource`:

```hcl
resource "null_resource" "provisioner" {
  provisioner "local-exec" {
    interpreter = ["/bin/bash", "-c"]
    command     = "<path_to>/ssm-provisioner.sh"
    environment = {
      SCRIPT      = "<path_to>/<script_name>"
      USERNAME    = "<username>"
      INSTANCE_ID = "<instance_id>"
      AWS_REGION  = "<aws_region>"
      PRIVATE_KEY = nonsensitive("<private_key>")
    }
  }
}
```

## Inputs

All inputs are provided as environment variables.

|Name|Required|Type|Default|Valid Options|Description|
|---|---|---|---|---|---|
|SCRIPT|yes|string|N/A|N/A|The local path of the script to execute on the instance.|
|USERNAME|yes|string|N/A|N/A|The username to use when connecting to the instance.|
|INSTANCE_ID|yes|string|N/A|N/A|The ID of the instance to connect to.|
|AWS_REGION|yes|string|N/A|N/A|The AWS region to use when connecting to the instance.|
|PRIVATE_KEY|yes|string|N/A|N/A|The private key content to use when connecting to the instance. The use of the `nonsensitive` function is suggested, otherwise the provisioner output will be hidden.|
|ENVIRONMENT|no|string|N/A|N/A|A string containing any environment variables to be set when executing the script. The use of a multi-line `EOF` string is suggested for ease of formatting, but is not required.|
|SSH_COMMAND|no|string|`bash -s`|N/A|The SSH command to use when executing the script.|
|SSH_PORT|no|number|`22`|N/A|The SSH port to use when connecting to the instance.|
|SSH_TIMEOUT|no|number|`120`|N/A|The SSH timeout to use when connecting to the instance, in seconds.|
|SSM_TIMEOUT|no|number|`120`|N/A|The timeout to use when waiting for SSM connection on the instance, in seconds.|
|SSM_SLEEP|no|number|`1`|N/A|The time to sleep when waiting for SSM connection on the instance, in seconds.|
|DEBUG|no|bool|`false`|`true`, `false`|Enable bash debugging via `set -x`, sensitive output will be still be hidden.|

## Examples

Useful examples of working Terraform modules are provided within the [modules folder](./modules).

## Releases

You can browse the latest releases in the [tags page](https://gitlab.com/zdzielinski/ssm-provisioner/-/tags), with more details in the [changelog page](./CHANGELOG.md).

## Contributing

If you'd like to help out, please read the [contributing page](./CONTRIBUTING.md).

## References

You can find additional reference documentation in the [references page](./docs/REFERENCES.md).

## License

This project uses [the MIT license](./LICENSE.md).

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.4](https://gitlab.com/zdzielinski/ssm-provisioner/compare/v0.0.3...v0.0.4) (2022-01-31)

### [0.0.3](https://gitlab.com/zdzielinski/ssm-provisioner/compare/v0.0.2...v0.0.3) (2022-01-31)


### Features

* added pre-commit hook via husky to check for repository secrets ([8f1c1ba](https://gitlab.com/zdzielinski/ssm-provisioner/commit/8f1c1baae8247fccf0c2fe83fd53e23c2f18e9b0))
* added simple npm test framework for ease of release validation ([28e0a4a](https://gitlab.com/zdzielinski/ssm-provisioner/commit/28e0a4a71f7d082705d43e8426c947df1d8fd5df))
* added the ability to pass environment variables to the remote script ([1988c12](https://gitlab.com/zdzielinski/ssm-provisioner/commit/1988c1262cd2515279fa32631ed606458116527c))


### Bug Fixes

* fixed bug where SSM timeout waiting was not working correctly, and could potentially never exit ([053b54c](https://gitlab.com/zdzielinski/ssm-provisioner/commit/053b54cbca81442258d06b1ad69286e0d63948e3))
* fixed bug where the SSH port was not being ingested correctly ([1ed46bd](https://gitlab.com/zdzielinski/ssm-provisioner/commit/1ed46bdc882c454f6c72f5e5c76c924b61dd87ec))

### [0.0.2](https://gitlab.com/zdzielinski/ssm-provisioner/compare/v0.0.1...v0.0.2) (2022-01-30)


### Bug Fixes

* fixed bug with bash debugging not re-enabling for one instance of sensitive output masking ([5d990cf](https://gitlab.com/zdzielinski/ssm-provisioner/commit/5d990cf15e1846e64fb00ec752e6d206d9a55a51))

### 0.0.1 (2022-01-30)


tests/test.js
const execSync = require('child_process').execSync;

// list of modules to test
const modules = [
  'modules/simple',
];

// run a command test with various printing options
function runCmdTest (header, command, timeout=0) {
  var success = true;
  console.log('-'.repeat(80));
  console.log(header);
  console.log(`> ${command}`)
  try {
    execSync(command, {
      timeout: timeout,
      stdio: 'inherit',
      shell: true
    });
  } catch (error) { success = false; }
  console.log(`success: ${success}`)
  return success;
}

// run tests against all modules
async function runTests () {
  // initial overall status is true
  var overallStatus = true;
  // loop over all modules
  for (const module of modules) {
    // clean module (timeout: default)
    if (!runCmdTest(`cleaning: ${module}`, 
      `rm -rf ${module}/.terraform ${module}/terraform.tfstate*`)) {
      overallStatus = false; continue; // fail and continue to next module
    }
    // init module (timeout: default)
    if (!runCmdTest(`testing - terraform init: ${module}`, 
      `terraform -chdir=${module} init -input=false -compact-warnings`)) {
      overallStatus = false; continue; // fail and continue to next module
    }
    // validate module (timeout: default)
    if (!runCmdTest(`testing - terraform validate: ${module}`, 
      `terraform -chdir=${module} validate -compact-warnings`)) {
      overallStatus = false; continue; // fail and continue to next module
    }
    // apply module (timeout: 120 seconds)
    if (!runCmdTest(`-- testing - terraform apply: ${module}`, 
      `terraform -chdir=${module} apply -auto-approve -input=false -compact-warnings`)) {
      overallStatus = false; // fail and continue within this module
    }
    // destroy module (timeout: default)
    if (!runCmdTest(`-- testing - terraform destroy: ${module}`, 
      `terraform -chdir=${module} destroy -auto-approve -input=false -compact-warnings`)) {
      overallStatus = false; // fail and continue within this module
    }
  }
  // return overall status
  return overallStatus;
}

// run tests and exit in error if needed
if (!runTests()) process.exit(1);

ssm-provisioner.sh
#!/usr/bin/env bash

# ssm-provisioner.sh
# https://gitlab.com/zdzielinski/ssm-provisioner
# v0.0.4

# The MIT License
# Copyright 2022 Zachariah Dzielinski
# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:
# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

# exit on error
set -e

# enable debug mode if specified
DEBUG="${DEBUG:-false}"
[[ "$DEBUG" = "true" ]] && set -x

# defaulted variables
SSH_COMMAND="${SSH_COMMAND:-"bash -s"}"
SSH_PORT="${SSH_PORT:-22}"
SSH_TIMEOUT="${SSH_TIMEOUT:-120}"
SSM_TIMEOUT="${SSM_TIMEOUT:-120}"
SSM_SLEEP="${SSM_SLEEP:-1}"

# verify client tools are installed
which aws >/dev/null || { >&2 echo "Error: The `aws` command is not installed"; exit 1; }

# verify existence of required variables
[ -z "${SCRIPT}" ] && { >&2 echo "Error: The `SCRIPT` environment variable is not set"; exit 1; }
[ -z "${USERNAME}" ] && { >&2 echo "Error: The `USERNAME` environment variable is not set"; exit 1; }
[ -z "${INSTANCE_ID}" ] && { >&2 echo "Error: The `INSTANCE_ID` environment variable is not set"; exit 1; }
[ -z "${AWS_REGION}" ] && { >&2 echo "Error: The `AWS_REGION` environment variable is not set"; exit 1; }
[[ "$DEBUG" = "true" ]] && set +x # hide sensitive content when debugging
[ -z "${PRIVATE_KEY}" ] && { >&2 echo "Error: The `PRIVATE_KEY` environment variable is not set"; exit 1; }
[[ "$DEBUG" = "true" ]] && set -x # reset into debug mode afterwards

# verify script path exists as a file
test -f $SCRIPT || { >&2 echo "Error: Could not find script `$SCRIPT`"; exit 1; }

# random tag name
RANDOM_TAG=$(printf "%04x" $RANDOM $RANDOM $RANDOM $RANDOM)

# temporary ssh config with random tag name
TMP_SSH_CONFIG=/tmp/$RANDOM_TAG.ssm_provisioner_ssh_config

cleanup () {
  # store original code
  ARG=$?
  # clean the temporary ssh config
  rm -f $TMP_SSH_CONFIG
  # exit with original code
  exit $ARG
}

# run cleanup on exit
trap cleanup EXIT

# clean and create temporary ssh config
rm -f $TMP_SSH_CONFIG
cat <<EOF > $TMP_SSH_CONFIG
host i-* mi-*
  StrictHostKeyChecking no
  UserKnownHostsFile /dev/null
  ProxyCommand sh -c "aws --region $AWS_REGION ssm start-session --target %h --document-name AWS-StartSSHSession --parameters 'portNumber=%p'"
EOF

# wait until ssm is available on the instance
NEXT_WAIT_TIME=0
until [ $NEXT_WAIT_TIME -ge $SSM_TIMEOUT ] || [ "$SSM_STATUS" = "connected" ]; do
  sleep $SSM_SLEEP
  NEXT_WAIT_TIME=$((NEXT_WAIT_TIME+SSM_SLEEP))
  SSM_STATUS=$(aws ssm get-connection-status --target $INSTANCE_ID --query Status --output text)
done

# error out if ssm could not connect before timeout
[ "$SSM_STATUS" != "connected" ] && { >&2 echo "Error: Timeout waiting for SSM to become connected on instance `$INSTANCE_ID`"; exit 1; }

# ingest the private key
eval $(ssh-agent) >/dev/null
[[ "$DEBUG" = "true" ]] && set +x # hide sensitive content when debugging
echo "$PRIVATE_KEY" | tr -d '\r' | ssh-add - >/dev/null
[[ "$DEBUG" = "true" ]] && set -x # reset into debug mode afterwards

# ingest and execute the script
ssh -F $TMP_SSH_CONFIG \
  -p $SSH_PORT \
  -o ConnectTimeout=$SSH_TIMEOUT \
  $USERNAME@$INSTANCE_ID \
  $ENVIRONMENT $SSH_COMMAND < $SCRIPT